specs/dwritecore/DWriteCore.md

```diff
@@ -78,33 +78,275 @@ with interoperability issues.
+    stretch, and style parameters.
+
+The compatibility issue with older fonts arises because the concept of typographic family name predates
+the concept of font axis values, which were introduced along with variable fonts in OpenType 1.8. Before
+OpenType 1.8, the typographic family name merely expressed the designer's intent that a set fonts were
+related, but with no guarantee that those fonts could be programmatically differentiated based on their
+properties. As a hypothetical example, suppose all of the following fonts have the typographic family
+name "Legacy":
+
+Full Name               | WWS Family   | Weight    | Stretch   | Style  | Typo Family   | wght  | wdth  | ital  | slnt
+------------------------|--------------|-----------|-----------|--------|---------------|-------|-------|-------|------
+Legacy                  | Legacy       | 400       | 5         | 0      | Legacy        | 400   | 100   | 0     | 0
+Legacy Bold             | Legacy       | 700       | 5         | 0      | Legacy        | 700   | 100   | 0     | 0
+Legacy Black            | Legacy       | 900       | 5         | 0      | Legacy        | 900   | 100   | 0     | 0
+Legacy Soft             | Legacy Soft  | 400       | 5         | 0      | Legacy        | 400   | 100   | 0     | 0
+Legacy Soft Bold        | Legacy Soft  | 700       | 5         | 0      | Legacy        | 700   | 100   | 0     | 0
+Legacy Soft Black       | Legacy Soft  | 900       | 5         | 0      | Legacy        | 900   | 100   | 0     | 0
+
+The "Legacy" typographic family has three weights and each weight has regular and "Soft" variants.
+If these were new fonts, they could be implemented as declaring a SOFT design axis. However, these fonts
+predate OpenType 1.8, so their only design axes are those derived from weight, stretch, and style. For
+each weight, this font family has two fonts with identical axis values, so it is not possible to 
+unambiguously select a font in this family using axis values alone.
+
+The hybrid font selection model mitigates this by allowing the specified family name to be _either_
+a typographic family name or a WWS family name. (It can also be an RBIZ name or full name.) In the
+example above, an application can avoid ambiguity by specifying "Legacy" or "Legacy Soft" as the
+family name. These two cases are handled as follows:
+
+  - "Legacy Soft" does not match a typographic family name, so only fonts in the WWS family
+    are considered. These can be differentiated by weight, so there is no ambiguity.
+
+  - "Legacy" matches both a typographic family and a WWS family. All fonts in the typographic
+    family are considered, but presence in the WWS family is used as a tie-breaker. For example,
+    Legacy Bold has the same weight as Legacy Soft Bold, but the former is preferred because
+    it is in the "Legacy" WWS family.
+
+Document and application compatibility issues are mitigated by allowing different types of font
```

Comment:
Do I understand correctly that this adaptation internally relies on string processing of the family name, i.e. detecting the "Soft" or "Black" suffix etc.?

Reply:
String processing is sometimes used to derive the weight-stretch-style family name, but it depends on the font.

For newer fonts with a STAT (style attributes) table, there are rules in OpenType 1.8 for how to construct family names and face names. A STAT table may be present in either a static or variable font, but is required only for variable fonts.

For static fonts without a STAT table, name ID 1 is the family name (what I've referred to here as the RBIZ family name), name ID 16 is the preferred/typographic family name, and name ID 22 is the WWS family name. If neither name 16 or 17 is present then all three family names are the same. If there's a name ID 16 but no 22 then a font may set a flag declaring itself to be a "WWS-only" family, in which case name ID 16 is used for both the typographic family name and the WWS family name.

Otherwise, if there's a name ID 16 but no 22 and no WWS-only flag, then DWrite uses a complicated heuristic to derived the WWS family name. This heuristic was first used by Windows Presentation Foundation (WPF) and later adopted by DWrite, and is described in a white paper called "WPF Font Selection Model" by Mikhail Leonov and David Brown. That white paper also describes the algorithm for ranking matching fonts based on weight, stretch, and style.